libs/core/execution_base/include/hpx/execution_base/any_sender.hpp

```diff
@@ -0,0 +1,769 @@
+        //   alignment_size alignment.
+        // - heap_storage: A pointer to base_type that is used when objects
+        //   don't fit in the embedded storage.
+        union
+        {
+            std::aligned_storage_t<embedded_storage_size, alignment_size>
+                embedded_storage;
+            base_type* heap_storage;
+        };
+        base_type* object;
+
+        // Returns true when it's safe to use the embedded storage, i.e.
+        // when the size and alignment of Impl are small enough.
+        template <typename Impl>
+        static constexpr bool can_use_embedded_storage()
+        {
+            constexpr bool fits_storage =
+                sizeof(std::decay_t<Impl>) <= embedded_storage_size;
+            constexpr bool sufficiently_aligned =
+                alignof(std::decay_t<Impl>) <= alignment_size;
+            return fits_storage && sufficiently_aligned;
+        }
+
+        bool using_embedded_storage() const noexcept
+        {
+            return object ==
+                reinterpret_cast<base_type const*>(&embedded_storage);
+        }
+
+        bool empty() const noexcept
+        {
+            return get().empty();
+        }
+
+        void release()
+        {
+            HPX_ASSERT(!empty());
+
+            if (using_embedded_storage())
+            {
+                get().~base_type();
+            }
+            else
+            {
+                delete heap_storage;
+                heap_storage = nullptr;
+            }
+
+            object = get_empty_vtable<base_type>::call();
+        }
+
+    public:
+        sbo_storage()
+          : heap_storage(nullptr)
+          , object(get_empty_vtable<base_type>::call())
+        {
+        }
+
+        ~sbo_storage()
+        {
+            if (!empty())
+            {
+                release();
+            }
+        }
+
+        sbo_storage(sbo_storage&& other)
+          : heap_storage(nullptr)
+          , object(get_empty_vtable<base_type>::call())
+        {
+            *this = std::move(other);
+        }
+
+        sbo_storage& operator=(sbo_storage&& other)
+        {
+            if (&other != this)
+            {
+                if (!empty())
+                {
+                    release();
+                }
+
+                if (!other.empty())
+                {
+                    if (other.using_embedded_storage())
+                    {
+                        auto p =
+                            reinterpret_cast<base_type*>(&embedded_storage);
+                        auto otherp = reinterpret_cast<base_type*>(
+                            &other.embedded_storage);
+                        // We cast the pointers to void* to silence warnings
+                        // about copying a dynamic class (which don't apply
+                        // here)
+                        std::memcpy(static_cast<void*>(p),
+                            static_cast<void*>(otherp), embedded_storage_size);
```

Comment:
Uhhh. This assumes that the embedded object is bitwise moveable. Shouldn't we explicitly `std::move` the rhs here?

Reply:
My understanding is that this is in fact ok, since we're managing the construction and destruction manually, but I'm definitely not 100% sure. If it's not ok, I'd again like to understand the difference to the polymorphic executor ;) What are the alternatives?

```
*p = std::move(*otherp)
```
without `p` holding a necessarily valid object seems problematic, and
```
new (p) {};
*p = std::move(*otherp);
```
requires a default constructor (though if that's required for this to be correct then so be it).

Reply:
What about `new (p) Impl(std::move(*otherp));`?

The code you have will probably work for 99.9% of all types. It will however break for instance for types that have a member pointing to another member of the same instance.

Reply:
Good, thanks for explaining! That's a corner case I did not consider. I can anyway use memcpy in the cases that the type is trivially copyable.

Reply:
I'm now placement-new-ing this. I'm leaving the memcpy for a maybe-sometime-in-the-future optimization.